src/Layer/LabelLayer.js

```diff
@@ -1,17 +1,134 @@
 import * as THREE from 'three';
 import LayerUpdateState from 'Layer/LayerUpdateState';
 import ObjectRemovalHelper from 'Process/ObjectRemovalHelper';
-import Layer from 'Layer/Layer';
+import GeometryLayer from 'Layer/GeometryLayer';
 import Coordinates from 'Core/Geographic/Coordinates';
 import Extent from 'Core/Geographic/Extent';
 import Label from 'Core/Label';
 import { FEATURE_TYPES } from 'Core/Feature';
 import { readExpression } from 'Core/Style';
+import { ScreenGrid } from 'Renderer/Label2DRenderer';
 
 const coord = new Coordinates('EPSG:4326', 0, 0, 0);
 
 const _extent = new Extent('EPSG:4326', 0, 0, 0, 0);
 
+const nodeDimensions = new THREE.Vector2();
+const westNorthNode = new THREE.Vector2();
+const labelPosition = new THREE.Vector2();
+
+/**
+ * DomNode is node of tree data structure for labels div.
+ *
+ * @class DomNode
+ */
+class DomNode {
+    #domVisibility = false;
+
+    constructor() {
+        this.dom = document.createElement('div');
+
+        this.dom.style.display = 'none';
+
+        this.visible = true;
+    }
+
+    get visible() { return this.#domVisibility; }
+
+    set visible(v) {
+        if (v !== this.#domVisibility) {
+            this.#domVisibility = v;
+            this.dom.style.display = v ? 'block' : 'none';
+        }
+    }
+
+    hide() { this.visible = false; }
+
+    show() { this.visible = true; }
+
+    add(node) {
+        this.dom.append(node.dom);
+    }
+}
+
+/**
+ * LabelsNode is node of tree data structure for LabelLayer.
+ * the node is made of dom elements and 3D labels.
+ *
+ * @class      LabelsNode
+ */
+class LabelsNode extends THREE.Group {
+    constructor(node) {
+        super();
+        // attached node parent
+        this.nodeParent = node;
+        // When this is set, it calculates the position in that frame and resets this property to false.
+        this.needsUpdate = true;
+    }
+
+    // instanciate dom elements
+    initializeDom() {
+        // create root dom
+        this.domElements = new DomNode();
+        // create labels container dom
+        this.domElements.labels = new DomNode();
+
+        this.domElements.add(this.domElements.labels);
+
+        this.domElements.labels.dom.style.opacity = '0';
+    }
+
+    // add node label
+    // add label 3d and dom label
+    addLabel(label) {
+        // add 3d object
+        this.add(label);
+
+        // add dom label
+        this.domElements.labels.dom.append(label.content);
+
+        // Batch update the dimensions of labels all at once to avoid
+        // redraw for at least this tile.
+        label.initDimensions();
+
+        // add horizon culling point if it's necessary
+        if (this.nodeParent.layer.isGlobeLayer && this.nodeParent.level < 4) {
```

Comment:
could you document this magic number and declare it as a constant? why only do horizon culling in low levels?

Reply:
i add
```js
// the horizon culling is applied to nodes that trace the horizon which
// corresponds to the low zoom node, that's why the culling is done for a zoom lower than 4.
```